source/service.go

```diff
@@ -559,18 +553,37 @@ func (sc *serviceSource) extractNodePortTargets(svc *v1.Service) (endpoint.Targe
 	return internalIPs, nil
 }
 
-func (sc *serviceSource) extractNodePortEndpoints(svc *v1.Service, nodeTargets endpoint.Targets, hostname string, ttl endpoint.TTL) []*endpoint.Endpoint {
+// Try to extract SRV records for all services with named ports
+func (sc *serviceSource) extractSrvRecordEndpoints(svc *v1.Service, hostname string, ttl endpoint.TTL, targetHosts ...string) []*endpoint.Endpoint {
 	var endpoints []*endpoint.Endpoint
 
 	for _, port := range svc.Spec.Ports {
-		if port.NodePort > 0 {
-			// build a target with a priority of 0, weight of 0, and pointing the given port on the given host
-			target := fmt.Sprintf("0 50 %d %s", port.NodePort, hostname)
-
-			// figure out the portname
-			portName := port.Name
-			if portName == "" {
-				portName = fmt.Sprintf("%d", port.NodePort)
+		if port.Name == "" {
+			// Unnamed ports do not have a SRV record.
+			continue
+		}
+
+		var exposedPort int32
+		switch svc.Spec.Type {
+		case v1.ServiceTypeLoadBalancer:
+			fallthrough
+		case v1.ServiceTypeNodePort:
+			exposedPort = port.NodePort
```

Comment:
I am afraid this won't work for `NodePort` type. K8s uses unnamed ports when exposing `NodePort` services.

> If you set the type field to NodePort, the Kubernetes control plane allocates a port from a range specified by --service-node-port-range flag (default: 30000-32767).

https://kubernetes.io/docs/concepts/services-networking/service/#nodeport